Fix caldav TZ interpretation of all day events

franc6 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->



## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
In the caldav component, all-day events are treated as if they are from 00:00 to 00:00 in UTC, instead of in the local timezone.  This code fixes the logic which interprets date-only events as being that date in UTC to being that date in local time.  This is similar to how datetimes without a timezone are treated.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #25814
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

Note: There's already a test for this change, but it was passing because the tests set the timezone to UTC and there's no good way to change it on a per-test basis (changing the global DEFAULT_TIME_ZONE is a bad idea since the tests aren't synchronous, and the patch decorator doesn't work because await is used for the component setup which references DEFAULT_TIME_ZONE and causes failure).  Several tests fail (before and after this PR) if the tests are forced to run in another timezone (by modifying homeassistant/util/dt.py).  However, the all-day tests will pass when the timezone is not UTC (after this PR).

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

franc6 committed:
Fixed #25814 by correcting TZ interpretation of all day of events

homeassistant commented:
Hi @franc6, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#48642).

Once you do that we will be able to review and accept this pull request.

Thanks!

franc6 commented:
Is there something still missing from this PR?  The bot wants to remove it because there's been no activity for a while.

balloob commented:
@franc6 can you add a test to make sure it works now.

franc6 commented:
Probably, it will depend on how much time I have. My own preference would be for the existing tests to run in multiple time zones every time. Having written the code, I feel confident it works in all time zones, but if I hadn’t written it, I’d want to see it tested in at least UTC, UTC + <some value>, and UTC - <some value> as a minimum. To be sure there’s no “magic time zone” that lets it work — as is the case without the patch, the code works well, if it’s running in UTC. :)

> On Jun 7, 2021, at 1:36 PM, Paulus Schoutsen ***@***.***> wrote:
> 
> ﻿
> @franc6 can you add a test to make sure it works now.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

franc6 commented:
A few notes to myself:
1) The tests are running synchronously, so changing DEFAULT_TIME_ZONE isn't an issue.
2) It looks like the all-day tests were passing without the patch when forced to a different timezone, because the date being used was within the same day still, and I missed that earlier.
3) Because different time zones overlap with UTC differently, we need to change what dt.now is returning to check multiple time combinations, since the attributes won't include the timezone, and do get adjusted properly.  Only the state is affected, this PR fixes the state to be ON during the local day, instead of ON only during the UTC day.

franc6 committed:
Updated tests for #25814

Please notes that unit tests were inadequate in the past, or 28514
would have been caught.  The new tests set one of three time zones
explicitly for all day event tests, and the all day event tests have
been modified to patch the return value of homeassistant.util.dt.now to
be in the local time zone, instead of always UTC.  By having it return
times in the local time zone, the tests better match real-world usage --
homeassistant.util.dt.now wouldn't return a UTC time when a different
time zone is set.

The unit tests now check for a single occurrence all day event at the
beginning of the local day, in the middle of the local day, and at the
end of the local day.  Each check is run in three separate time zones,
Asia/Baghdad, UTC, and America/New_York.  The times are chose such that
the early in the local day time for Asia/Baghdad is actually the day
before in UTC, and the late in the local day time for America/New_York
is actually the next day in UTC.  This guarantees that when the tests
pass, there are no internal checks that shifted what should be a local
time as if it were UTC being shifted to local time, which was bug #25814.

Similar changes were made for the recurring all day event test, too.

Note that if you run the new unit tests without the fixed code in this
patch, that the following tests will fail:

test_all_day_event_returned_early[pyloop-baghdad]
test_all_day_event_returned_late[pyloop-new_york]
test_event_rrule_all_day_late[pyloop-new_york]

After applying the fix, all tests will succeed.  If you're wondering why
test_event_rrule_all_day_early[pyloop-baghdad] doesn't fail without the
fix, it's because we're not checking on the first occurrence of the
event.  If we did that, the test would fail just like
test_all_day_event_returned_early[pyloop-baghdad].  However, since by
checking the first occurrence we're running the same condition that
test_all_day_event_returned_early[pyloop-baghdad] already checks, it
would be redundant to add a test which fails for a recurring event in
that time zone.  The whole point of the test_event_rrule_all_day tests
is to ensure that recurrences are returned when it's not the first
occurrence.

franc6 commented:
OK, unit tests have been updated.  Please see the comment for the commit for the unit tests to explain the reasons for the changes.  I strongly recommend running the updated unit tests WITHOUT the fix, so the failures can be observed, then apply the fix, and observe the tests pass. :)
If you've been watching this conversation, please note that I updated my "notes to myself" to remove incorrect information.  Let me know if there are further questions or edits you'd like to see.
